### PR TITLE
feat(router): attach 'router-link-active' class automatically to all …

### LINKED ISSUE #9426: Router: Feature Request: default 'router-link-active' class attachment (old router behaviour)
**I'm submitting a ...** 
```
[ ] bug report
[x] feature request
[ ] support request
```

**Current behavior**

angular/router 3.0.0-alpha.7 introduced the `routerLinkActive` directive, which enabled us to set a custom class on the current `routerLink` element. This is great!

But AFAIK, I have to attach this routerLinkActive directive to EVERY SINGLE LINK.

eg

```
<ul>
    <li><a [routerLink]="['/sales']"    routerLinkActive="router-link-active" ...
    <li><a [routerLink]="['/reports']"  routerLinkActive="router-link-active" ...
    <li><a [routerLink]="['/search']"   routerLinkActive="router-link-active" ...
    <li><a [routerLink]="['/settings']" routerLinkActive="router-link-active" ...
    ...
</ul>
```

**Expected/desired behavior**

Angular attach the class `router-link-active` automatically to all elements with `routerLink` directives if `routerLinkActive` is not present (i.e. overridden).

Then, all the user has to include is a custom `.router-link-active { }` CSS style and they get active link feedback for free! (just like the old router)

Alternatively, add an option to set `routerLinkActive="custom-class"` once only? Per app ? Per router ? Per parent element (i.e. the `<ul>`)?

Maybe the `routerLinkActive` selector can be updated to match `routerLink` elements. Or through `[routerLinkActiveOptions]="{autoapply: true}"` at the `<ul>` level or similar ?

**What is the motivation / use case for changing the behavior?**

Better user experience. Just add one CSS style and get active link feedback for free. It was previous router behavior. Reduce template noise.

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.2   +  angular/router 3.0.0-alpha.7
 
* **Language:**  TypeScript 1.8.10

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/9426

### CODE REVIEW

**Comment by laco0416** [comment]
I think it'll be good to be able to configure the class name in `ExtraOptions`

```ts
RouterModule.forRoot(routes, {
  routerLinkActiveClass: 'my-link-active'
});
```

File: modules/@angular/router/src/directives/router_link_active.ts [hunk 1]
```diff
@@ -72,13 +74,13 @@ import {RouterLink, RouterLinkWithHref} from './router_link';
...
  *
- * @selector ':not(a)[routerLink]'
+ * @selector '[routerLinkActive],a[routerLink],:not(a)[routerLink]'
  * @ngModule RouterModule
  *
  * @stable
  */
 @Directive({
-  selector: '[routerLinkActive]',
+  selector: '[routerLinkActive],a[routerLink],:not(a)[routerLink]',
```
Comment (zoechi) [inline modules/@angular/router/src/directives/router_link_active.ts]: Wouldn't this do the same?
```
selector: '[routerLinkActive],[routerLink]',
```

Comment (DzmitryShylovich) [inline modules/@angular/router/src/directives/router_link_active.ts]: yeah, it doesn't matter in this case if it's an anchor or another element

**Comment by zoechi** [comment]
I didn't measure but calculating router-link-active seems rather expensive and It's common to have lots of router links on a page. I guess this is why it wasn't added by default. 
There also seems no way to opt out.

**Comment by laco0416** [comment]
good point. maybe it should apply the class only if given `activeClass`.

File: modules/@angular/router/src/directives/router_link_active.ts [hunk 2]
```diff
@@ -87,14 +89,27 @@ export class RouterLinkActive implements OnChanges,
...
+  }
 
-  constructor(private router: Router, private element: ElementRef, private renderer: Renderer) {
-    this.subscription = router.events.subscribe(s => {
-      if (s instanceof NavigationEnd) {
+  get routerLinkActiveOptions() { return this._routerLinkActiveOptions; }
+
+  constructor(
+      private router: Router, private element: ElementRef, private renderer: Renderer,
+      @Inject(ROUTER_CONFIGURATION) @Optional() private config: any /* TODO #12631 ExtraOptions*/) {
```
Comment (vsavkin) [inline modules/@angular/router/src/directives/router_link_active.ts]: Why is config optional?

**Comment by vsavkin** [comment]
Overall, it's good, but i have a few questions/suggestions.

1. Change the configuration to be something like this:

RouterModule.forRoot(ROUTES, {
  routerLinkActive: {
    defaultClass: 'someclass',
    exact: true
  }
});

2. Currently, the user has to opt in to enable routerLinkActive. Since it's provided by default, there should be a way to opt out.

3. Do we have to support router-level configuration? The router-link-active name is fairly unique, so I'm not sure who would need to change it. What do you think?

**Comment by DzmitryShylovich** [comment]
> Change the configuration to be something like this:

Good idea. Also was thinking about that.

> Do we have to support router-level configuration? The router-link-active name is fairly unique, so I'm not sure who would need to change it. What do you think?

I think we 100% need `routerLinkActive.exact` if we will enable routerLinkActive by default. About default class name it's better to ask @laco0416 it was his [idea](https://github.com/angular/angular/pull/12832#issuecomment-260158406) :)

> Currently, the user has to opt in to enable routerLinkActive. Since it's provided by default, there should be a way to opt out.

do u have an idea how it can be implemented?

File: modules/@angular/router/src/directives/router_link_active.ts (hunk 2 above)
Comment (DzmitryShylovich) [inline modules/@angular/router/src/directives/router_link_active.ts]: not to break existing tests :)

**Comment by DzmitryShylovich** [comment]
Updated.

**Comment by laco0416** [comment]
It seems a little compilcated behavior. I think it can be simpler and more understandable easily as a new feature. My expecting behavior is the below   

```
// no changes without routerLinkActive
<a routerLink=""></a>

// no changes with routerLinkActive and its value
<a routerLink="" routerLinkActive="my-active-link"></a> // => class="my-active-link"

// new fearure with value-less routerLinkActive 
<a routerLink="" routerLinkActive></a> // => class="routrr-link-active"
```

this design can keep current performance and customizable in router level. how do you think about this?

**Comment by laco0416** [comment]
Oh, I'm sorry. I'm misunderstanding current `routerLinkActive` spec.

**Pull request closed without merging**